src/3d/framegraph/qgsambientocclusionrenderview.cpp

```diff
@@ -0,0 +1,183 @@
+/***************************************************************************
+  qgsambientocclusionrenderview.cpp
+  --------------------------------------
+  Date                 : June 2024
+  Copyright            : (C) 2024 by Benoit De Mezzo and (C) 2020 by Belgacem Nedjima
+  Email                : benoit dot de dot mezzo at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsambientocclusionrenderview.h"
+#include <Qt3DRender/QCamera>
+#include <Qt3DRender/QLayerFilter>
+#include <Qt3DRender/QLayer>
+#include <Qt3DRender/QRenderTargetSelector>
+#include <Qt3DRender/QRenderTarget>
+#include <Qt3DRender/QTexture>
+#include <Qt3DRender/QClearBuffers>
+#include <Qt3DRender/qsubtreeenabler.h>
+#include <Qt3DRender/QCameraSelector>
+#include <Qt3DRender/QRenderStateSet>
+#include <Qt3DRender/QDepthTest>
+#include <Qt3DRender/QCullFace>
+
+
+QgsAmbientOcclusionRenderView::QgsAmbientOcclusionRenderView( const QString &viewName, Qt3DRender::QCamera *mainCamera, QSize mSize, Qt3DRender::QTexture2D *forwardDepthTexture, Qt3DCore::QEntity *rootSceneEntity )
+  : QgsAbstractRenderView( viewName )
+  , mMainCamera( mainCamera )
+{
+  mAOPassLayer = new Qt3DRender::QLayer;
+  mAOPassLayer->setRecursive( true );
+  mAOPassLayer->setObjectName( mViewName + "::Layer(AO)" );
+
+  mBlurPassLayer = new Qt3DRender::QLayer;
+  mBlurPassLayer->setRecursive( true );
+  mBlurPassLayer->setObjectName( mViewName + "::Layer(Blur)" );
+
+  // ambient occlusion rendering pass
+  buildRenderPass( mSize, forwardDepthTexture, rootSceneEntity );
+}
+
+void QgsAmbientOcclusionRenderView::updateWindowResize( int width, int height )
+{
+  mAOPassTexture->setSize( width, height );
+  mBlurPassTexture->setSize( width, height );
+}
+
+void QgsAmbientOcclusionRenderView::setEnabled( bool enable )
+{
+  QgsAbstractRenderView::setEnabled( enable );
+  if ( mAmbientOcclusionRenderEntity != nullptr )
+    mAmbientOcclusionRenderEntity->setEnabled( enable );
+  if ( mAmbientOcclusionBlurEntity != nullptr )
+    mAmbientOcclusionBlurEntity->setEnabled( enable );
+}
+
+Qt3DRender::QRenderTarget *QgsAmbientOcclusionRenderView::buildTextures( QSize mSize )
```

Comment:
it is confusing that buildTextures() creates one texture, and the other texture is created in buildRenderPass() - both should probably be created here

Reply:
indeed